crates/apps/src/lib/node/ledger/shell/finalize_block.rs

```diff
@@ -652,6 +519,293 @@ where
+                tx_event
+                    .extend(WithGasUsed(tx_result.gas_used))
+                    .extend(Info("Check batch for result.".to_string()))
+                    .extend(Batch(&tx_result.to_result_string()));
+            }
+            Err(Error::TxApply(protocol::Error::WrapperRunnerError(msg))) => {
+                tracing::info!(
+                    "Wrapper transaction {} failed with: {}",
+                    tx_event["hash"],
+                    msg,
+                );
+                tx_event
+                    .extend(WithGasUsed(consumed_gas))
+                    .extend(Info(msg.to_string()))
+                    .extend(Code(ResultCode::InvalidTx));
+                // Make sure to clean the write logs for the next transaction
+                self.state.write_log_mut().drop_tx();
+                self.state.write_log_mut().drop_batch();
+            }
+            Err(msg) => {
+                // This branch represents an error that affects the entire
+                // batch
+                tracing::info!(
+                    "Transaction {} failed with: {}",
+                    tx_event["hash"],
+                    msg
+                );
+                // TODO: instead of dropping everything, commit the successful
+                // txs before the error (and emit their events) and adjust the
+                // replay protection handling
+
+                // If user transaction didn't fail because of out of gas nor
+                // invalid section commitment, commit its hash to prevent
+                // replays
+                if matches!(tx_header.tx_type, TxType::Wrapper(_)) {
+                    if !matches!(
+                        msg,
+                        Error::TxApply(protocol::Error::GasError(_))
+                            | Error::TxApply(protocol::Error::ReplayAttempt(_))
+                    ) {
+                        self.commit_batch_hash(replay_protection_hashes);
+                    } else if let Error::TxApply(
+                        protocol::Error::ReplayAttempt(_),
+                    ) = msg
+                    {
+                        // Remove the wrapper hash but keep the inner tx
+                        // hash. A replay of the wrapper is impossible since
+                        // the inner tx hash is committed to storage and
+                        // we validate the wrapper against that hash too
+                        let header_hash = replay_protection_hashes
+                            .expect("This cannot fail")
+                            .header_hash;
+                        self.state
+                            .redundant_tx_hash(&header_hash)
+                            .expect("Error while marking tx hash as redundant");
+                    }
+                }
+
+                for _ in 0..commitments_len {
+                    stats.increment_errored_txs();
+                }
```

Comment:
```suggestion
                stats.increment_errored_txs_by(commitments_len);                
```
something like this to avoid looping